lib/elixir/src/elixir_json.erl

```diff
@@ -304,6 +304,8 @@ encode_list(List, Encode) when is_list(List) ->
 
 do_encode_list([], _Encode) ->
     <<"[]">>;
+do_encode_list([{Key, _Value} | _Rest] = List, Encode) when is_atom(Key) ->
+    encode_key_value_list_checked(List, Encode);
```

Comment:
One edge case that could be done either way is encoding an invalid keyword list list such as `[k1: "a", k2: "b", "BAD"]`, which isn't valid syntax in that form, but could still be expressed as `[{:k1, "a", {:k2, "b"}, "BAD"]`.

This is using `encode_key_value_list_checked/2`, which will correctly raise an error on bad input: if there is a dup key, or any element that doesn't match `{Key, Value}` in the list, but will have a performance overhead.

Alternative is to just use `encode_key_value_list/2`, which is faster, but will allow duplicate keys to be encoded, and will just skip invalid list items.